R/clone_package.R

```diff
@@ -0,0 +1,285 @@
+#' another.
+#'
+#' @param mn (MNode/CNode) The Node to query for Object sizes
+#' @param resource_map_pid (character) The identifier of the Data Package's Resource Map
+#' @param formatType (character) Optional. Filter to just Objects of the given
+#' formatType. One of METADATA, RESOURCE, or DATA or * for all types
+#'
+#' @author Dominic Mullen, \email{dmullen17@@gmail.com}
+#'
+#' @return (character) The formatId, fileName, and identifier of each data object in a package.
+get_package_metadata <- function(mn, resource_map_pid, formatType = "DATA") {
+    query <- dataone::query(mn,
+                            paste0("q=resourceMap:\"",
+                                   resource_map_pid,
+                                   "\"+AND+formatType:",
+                                   formatType,
+                                   "&fl=formatId+AND+fileName+AND+identifier"),
+                            as = "data.frame")
+
+    if (nrow(query) == 0) {
+        return(0)
+    }
+
+    # Replace NA fields
+    query$fileName[which(is.na(query$fileName))] <- query$identifier
+
+    return(query)
+}
+
+#' Return data identifiers from within an EML
+#'
+#' This function returns data object identifers present within an EML (electronic
+#' metadata language) document.
+#'
+#' @param eml () an EML object
+#'
+#' @return
+#' Returns a list of data object pids in the eml
+get_eml_pids <- function(eml) {
+    # message if no dataTables / otherEntites are found
+
+    # Get urls from EML
+    dataTable_urls <- EML::eml_get(eml@dataset@dataTable, "url")
+    otherEntity_urls <- EML::eml_get(eml@dataset@otherEntity, "url")
+
+    urls <- vector("character")
+    urls <- c(urls, unlist(dataTable_urls), unlist(otherEntity_urls))
+
+    # Select the characters after the last "/" (the pid)
+    pids <- sapply(urls, function(url) {
+        split_string <- strsplit(url, split = "\\/")
+        utils::tail(split_string[[1]], n = 1)
+    })
+
+    return(pids)
+}
+
+#' Clone data pids between Dataone Member Nodes.
+#'
+#' This is a helper function for datamgmt::clone_package
+#'
+#' @param resource_map_pid (character) Resource map for the data pids.
+#' @param data_pids (character) Data object pids.
+#' @param from (MNode) Dataone Member Node to clone data objects from
+#' @param to (MNode) Dataone Member Node to clone data objects to
+#' @param n_max (integer) Optional. Number of tries to download a DataOne object.
+#' Can fail due to internet connectivity, suggested n_max >= 3.
+#'
+#' @return (list) Vector of data object pids.
+clone_data_objects <- function(resource_map_pid,
+                               data_pids,
+                               from,
+                               to,
+                               n_max = 3L) {
+    stopifnot(is.character(resource_map_pid))
+    stopifnot(is.character(data_pids))
+    stopifnot(length(data_pids) > 0)
+    stopifnot(methods::is(from, "MNode"))
```

Comment:
not sure if you have to explicitly reference `methods::` here. Fine to do so though

Reply:
Travis might've gotten annoyed at that - not certain though.